packages/block-editor/src/components/inserter/hooks/use-insertion-point.js

```diff
@@ -81,9 +81,9 @@ function useInsertionPoint( {
 			return getBlockIndex( clientId, destinationRootClientId );
 		}
 
-		// If there a selected block, we insert after the selected block.
+		// If there's a selected block, and the selected block is not the destination root block, we insert after the selected block.
 		const end = getBlockSelectionEnd();
-		if ( ! isAppender && end ) {
+		if ( ! isAppender && end && end !== destinationRootClientId ) {
```

Comment:
I'm not sure I understand the situation where `end` would be equal to `destinationRootClientId`, since `end` is usually representative of the last selected block and `destinationRootClientId` would be representative of that block's parent.

Reply:
I think it possibly indicates that something has gone wrong up here if the two could ever have the same value:
https://github.com/WordPress/gutenberg/blob/a58272e2ee19358321c01dbce7263fa5e27ad6a0/packages/block-editor/src/components/inserter/hooks/use-insertion-point.js#L50-L56

Reply:
Hmmm interesting...

The code above only happens when there's no `rootClientId` present. However, in the widgets screen, there's a `rootClientId` provided.

In the widget screen case, `rootClientId` is the last selected `widget-area` block, which is also the _selection end_ block. Since the `widget-area` block is a root block, there's no parent block for it.

I'm also hesitate that this fix could potentially break other usages in the post editor or other editor, but I couldn't find any other elegant solution. Suggestions welcome 🙇‍♂️!

Reply:
> In the widget screen case, rootClientId is the last selected widget-area block, which is also the selection end block

I don't think that can be the case, as it'd make the next line invalid:
```js
getBlockIndex( end, destinationRootClientId );
```

The arguments are `clientId` and `rootClientId`. Since `rootClientId` represents the parent of the `clientId`, if they're the same it'd mean the block is a child of itself.

I think there must be something else happening here causing the wrong clientId and rootClientId to be set within this function. Let me see if I can find some clues.

Reply:
I think it's this bit here @kevin940726:
https://github.com/WordPress/gutenberg/blob/643007e68633388280c9fc4456081fa5774cc45a/packages/edit-widgets/src/hooks/use-last-selected-root-id.js#L30

The rootClientId returned should ideally stay `null`.

Reply:
This is a bit more complex than I initially thought. I think the main issue is that the `Library` component doesn't really support the use case of widget areas.

A `widget-area` is really a special kind of block where when it's selected we don't want to insert after it, we want to insert inside of it.

The `useInsertionPoint` hook instead tries to insert after any selected block.

We can pass both a `rootClientId` and a `clientId` to `Library` to circumvent inserting after, but this always inserts before the given `clientId`, and doesn't cover the use case where the widget area has no inner blocks (and therefore no `clientId` can be passed).

Some options:
- Add an `insertionIndex` prop to `Library` so that we can override `useInsertionPoint`'s derived index and support inserting within widget areas when they're selected.
- Add some kind of support for container blocks where the default is to insert within them. I'm not sure if there are any other use cases for this kind of block.

Any other ideas?

Reply:
@kevin940726 I tired the `insertionIndex` idea in #25763